custom_components/opnsense/services.py

```
@@ -30,277 +31,576 @@
+    Raises:
+        ServiceValidationError: If no OPNsense clients are selected or the voucher server fails.
+    """
+    if not clients:
+        raise _service_validation_error(_TRANSLATION_KEY_GENERATE_VOUCHERS_FAILED)
+
+    voucher_list: list[Any] = []
+    for client in clients:
+        try:
+            vouchers: list[Any] = await client.generate_vouchers(call_data)
+        except OPNsenseVoucherServerError as e:
+            _LOGGER.error("Error getting vouchers from %s. %s", client.name, e)
```

Ruff (0.15.21)

[warning] Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)